File: src/frontend/cte/lib/app/data/model/cte_rodoviario_pedagio_model.dart
import 'dart:convert';
import 'package:pluto_grid/pluto_grid.dart';

import 'package:cte/app/infra/infra_imports.dart';
import 'package:cte/app/data/model/model_imports.dart';

class CteRodoviarioPedagioModel {
	int? id;
	int? idCteRodoviario;
	String? cnpjFornecedor;
	String? comprovanteCompra;
	String? cnpjResponsavel;
	double? valor;
	CteRodoviarioModel? cteRodoviarioModel;

	CteRodoviarioPedagioModel({
		this.id,
		this.idCteRodoviario,
		this.cnpjFornecedor,
		this.comprovanteCompra,
		this.cnpjResponsavel,
		this.valor,
		this.cteRodoviarioModel,
	});

	static List<String> dbColumns = <String>[
		'id',
		'cnpj_fornecedor',
		'comprovante_compra',
		'cnpj_responsavel',
		'valor',
	];
	
	static List<String> aliasColumns = <String>[
		'Id',
		'Cnpj Fornecedor',
		'Comprovante Compra',
		'Cnpj Responsavel',
		'Valor',
	];

	CteRodoviarioPedagioModel.fromJson(Map<String, dynamic> jsonData) {
		id = jsonData['id'];
		idCteRodoviario = jsonData['idCteRodoviario'];
		cnpjFornecedor = jsonData['cnpjFornecedor'];
		comprovanteCompra = jsonData['comprovanteCompra'];
		cnpjResponsavel = jsonData['cnpjResponsavel'];
		valor = jsonData['valor']?.toDouble();
		cteRodoviarioModel = jsonData['cteRodoviarioModel'] == null ? CteRodoviarioModel() : CteRodoviarioModel.fromJson(jsonData['cteRodoviarioModel']);
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonData = <String, dynamic>{};

		jsonData['id'] = id != 0 ? id : null;
		jsonData['idCteRodoviario'] = idCteRodoviario != 0 ? idCteRodoviario : null;
		jsonData['cnpjFornecedor'] = Util.removeMask(cnpjFornecedor);
		jsonData['comprovanteCompra'] = comprovanteCompra;
		jsonData['cnpjResponsavel'] = Util.removeMask(cnpjResponsavel);
		jsonData['valor'] = valor;
		jsonData['cteRodoviarioModel'] = cteRodoviarioModel?.toJson;
	
		return jsonData;
	}
	
	String objectEncodeJson() {
		final jsonData = toJson;
		return json.encode(jsonData);
	}

	plutoRowToObject(PlutoRow plutoRow) {
		id = plutoRow.cells['id']?.value;
		idCteRodoviario = plutoRow.cells['idCteRodoviario']?.value;
		cnpjFornecedor = plutoRow.cells['cnpjFornecedor']?.value;
		comprovanteCompra = plutoRow.cells['comprovanteCompra']?.value;
		cnpjResponsavel = plutoRow.cells['cnpjResponsavel']?.value;
		valor = plutoRow.cells['valor']?.value?.toDouble();
		cteRodoviarioModel = CteRodoviarioModel();
		cteRodoviarioModel?.rntrc = plutoRow.cells['cteRodoviarioModel']?.value;
	}	

	CteRodoviarioPedagioModel clone() {
		return CteRodoviarioPedagioModel(
			id: id,
			idCteRodoviario: idCteRodoviario,
			cnpjFornecedor: cnpjFornecedor,
			comprovanteCompra: comprovanteCompra,
			cnpjResponsavel: cnpjResponsavel,
			valor: valor,
		);			
	}

	
}

File: src/backend/cte/src/model/cte_rodoviario_pedagio_model.py
from src import db
from src.model.cte_rodoviario_model import CteRodoviarioModel


class CteRodoviarioPedagioModel(db.Model):
    __tablename__ = 'cte_rodoviario_pedagio'

    id = db.Column(db.Integer, primary_key=True)
    cnpj_fornecedor = db.Column(db.String(14))
    comprovante_compra = db.Column(db.String(20))
    cnpj_responsavel = db.Column(db.String(14))
    valor = db.Column(db.Float)
    id_cte_rodoviario = db.Column(db.Integer, db.ForeignKey('cte_rodoviario.id'))

    cte_rodoviario_model = db.relationship('CteRodoviarioModel', foreign_keys=[id_cte_rodoviario])

    def mapping(self, data):
        self.id = data.get('id')
        self.id_cte_rodoviario = data.get('idCteRodoviario')
        self.cnpj_fornecedor = data.get('cnpjFornecedor')
        self.comprovante_compra = data.get('comprovanteCompra')
        self.cnpj_responsavel = data.get('cnpjResponsavel')
        self.valor = data.get('valor')

    def serialize(self):
        return {
            'id': self.id,
            'idCteRodoviario': self.id_cte_rodoviario,
            'cnpjFornecedor': self.cnpj_fornecedor,
            'comprovanteCompra': self.comprovante_compra,
            'cnpjResponsavel': self.cnpj_responsavel,
            'valor': self.valor,
            'cteRodoviarioModel': self.cte_rodoviario_model.serialize() if self.cte_rodoviario_model else None,
        }